File: old-project/backend/models/event.go
package models

import (
	"gorm.io/gorm"
)

type Event struct {
	gorm.Model
	Title          string       `json:"title"`
	Description    string       `json:"description"`
	Date           string       `json:"date"`         // Can be a date, duration, or date range as string
	Link           string       `json:"link"`         // Learn more link
	SignUpLink     string       `json:"sign_up_link"` // Registration/signup form link
	OrganizationID uint         `json:"organization_id"`
	Organization   Organization `json:"organization,omitempty" gorm:"foreignKey:OrganizationID"`
}

// CreateEvent creates a new event
func CreateEvent(db *gorm.DB, event *Event) error {
	return db.Create(event).Error
}

// GetEventByID retrieves an event by ID
func GetEventByID(db *gorm.DB, id uint) (*Event, error) {
	var event Event
	err := db.Preload("Organization").First(&event, id).Error
	if err != nil {
		return nil, err
	}
	return &event, nil
}

// GetAllEvents retrieves all events
func GetAllEvents(db *gorm.DB) ([]Event, error) {
	var events []Event
	err := db.Preload("Organization").Order("created_at desc").Find(&events).Error
	return events, err
}

// GetEventsByOrganizationID retrieves all events for a specific organization
func GetEventsByOrganizationID(db *gorm.DB, orgID uint) ([]Event, error) {
	var events []Event
	err := db.Where("organization_id = ?", orgID).Order("created_at desc").Find(&events).Error
	return events, err
}

// UpdateEvent updates an existing event
func UpdateEvent(db *gorm.DB, event *Event) error {
	return db.Save(event).Error
}

// DeleteEvent deletes an event by ID
func DeleteEvent(db *gorm.DB, id uint) error {
	return db.Delete(&Event{}, id).Error
}

// DeleteEventByIDAndOrg deletes an event only if it belongs to the organization
func DeleteEventByIDAndOrg(db *gorm.DB, eventID, orgID uint) error {
	result := db.Where("id = ? AND organization_id = ?", eventID, orgID).Delete(&Event{})
	if result.RowsAffected == 0 {
		return gorm.ErrRecordNotFound
	}
	return result.Error
}

File: lib/services/event.ts
import prisma from "@/lib/prisma";

/**
 * Create a new event
 */
export async function createEvent(
  organizationId: bigint,
  title: string,
  description?: string,
  date?: string,
  link?: string,
  signUpLink?: string
) {
  const event = await prisma.event.create({
    data: {
      organizationId,
      title,
      description: description || null,
      date: date || null,
      link: link || null,
      signUpLink: signUpLink || null,
    },
    include: {
      organization: true,
    },
  });

  return event;
}

/**
 * Get event by ID
 */
export async function getEventById(id: bigint) {
  const event = await prisma.event.findUnique({
    where: { id },
    include: {
      organization: true,
    },
  });

  if (!event) {
    throw new Error("Event not found");
  }

  return event;
}

/**
 * Get all events
 */
export async function getAllEvents() {
  return prisma.event.findMany({
    include: {
      organization: true,
    },
    orderBy: { createdAt: "desc" },
  });
}

/**
 * Get events by organization ID
 */
export async function getEventsByOrganizationId(organizationId: bigint) {
  return prisma.event.findMany({
    where: { organizationId },
    include: {
      organization: true,
    },
    orderBy: { createdAt: "desc" },
  });
}

/**
 * Update an event
 */
export async function updateEvent(id: bigint, updates: any) {
  // Normalize incoming keys: support both camelCase and snake_case from the API
  const data: any = {};
  console.log("updateEvent incoming updates:", updates);
  if (updates.title !== undefined) data.title = updates.title;
  if (updates.description !== undefined) data.description = updates.description;
  if (updates.date !== undefined) data.date = updates.date;
  if (updates.link !== undefined) data.link = updates.link;
  // signUpLink may come as camelCase or snake_case (sign_up_link)
  if (updates.signUpLink !== undefined) data.signUpLink = updates.signUpLink;
  if (updates.sign_up_link !== undefined) data.signUpLink = updates.sign_up_link;
  console.log("updateEvent normalized data:", data);

  const event = await prisma.event.update({
    where: { id },
    data,
    include: {
      organization: true,
    },
  });

  return event;
}

/**
 * Delete an event
 */
export async function deleteEvent(id: bigint) {
  await prisma.event.delete({ where: { id } });
}

/**
 * Delete event only if it belongs to the organization
 */
export async function deleteEventByIdAndOrg(
  eventId: bigint,
  organizationId: bigint
) {
  const result = await prisma.event.deleteMany({
    where: {
      id: eventId,
      organizationId,
    },
  });

  if (result.count === 0) {
    throw new Error("Event not found or unauthorized");
  }
}
